_posts/2025-03-12-orange-telco-core-network-api-management-with-quarkus.adoc

```diff
@@ -0,0 +1,110 @@
+
+To achieve this goal, Orange needed a flexible and lightweight framework capable of handling constrained API exposure while ensuring compatibility with existing network core systems.
+
+After a rigorous evaluation of multiple frameworks and toolkits — including Quarkus, Ktor, Micronaut, and Vert.x — Orange selected Quarkus as the ideal solution.
+
+https://www.linkedin.com/in/patriceconil/[Patrice Conil], software craftsman at Orange, told us why.
+
+== Why Quarkus? A Framework Evaluation
+
+The team at Orange embarked on a comparative study to test alternatives to Spring Boot in a well-defined API wrapper exposure context.
+Key evaluation criteria included:
+
+- Learning Curve: How easily could a Spring developer transition?
+- Kotlin Compatibility: Could the framework work seamlessly with Kotlin?
+- Runtime footprint: Could it operate efficiently in a constrained environment?
+- Ease of Deployment: How smoothly could it be deployed on Kubernetes?
+
+After extensive testing, Quarkus stood out for multiple reasons:
+
+- Simplicity: A relatively simple learning curve, especially for those familiar with JAX-RS / Jakarta REST.
+- Dev Mode: Very fast startup, live reload and zero configuration Dev Services (Vault, Redis) result in great developer productivity.
+- Modularity: Only required dependencies were embedded, keeping applications lean.
+- Documentation: Well-organized, versioned documentation with working examples.
+- Native Compilation: The ability to generate compact native binaries for Kubernetes deployment.
+
+Despite the strong competition from a vibrant JVM ecosystem, these advantages made Quarkus the preferred choice for exposing 5G APIs at Orange.
+
+== Lessons Learned from Adopting Quarkus
+
+=== Migration & Development Experience
+
+Transitioning from Spring to Quarkus was not overly complex, especially for those familiar with JAX-RS / Jakarta REST.
+
+The https://quarkus.io/guides/dev-services[Dev Services] significantly enhanced productivity in dev and test modes, but required careful consideration for their CI/CD environments, where no container runtime is availabe.
+
+The modular approach ensured applications remained lightweight but introduced a large number of small dependencies to manage.
+
+=== Spec-First API Development
+
+As spec-first practitioners, integrating OpenAPI specification generation was a crucial requirement that Quarkus handled effectively.
+
+=== Java & Kotlin Interoperability
+
+While Quarkus supports Kotlin, writing full Kotlin extensions proved challenging at times.
+The team retained some Java code, and Java-Kotlin interoperability worked smoothly.
+
+=== Native Compilation & Performance
+
+Native compilation produced compact native executables, but the process of producing them was lengthy.
+The team reserved it for the final build stage when absolutely necessary.
+
+Some fine-tuning was required to prevent class pruning issues.
```

Comment:
Probably worth explaining this for people not familiar with GraalVM.

Reply:
Could you suggest a link to the doc for this? I have a (good?) idea of what Patrice meant, but I must admit I don't feel comfortable explaining it.

Reply:
Thank you Guillaume!